docs/docs/debugging-incremental-builds.md

```diff
@@ -0,0 +1,87 @@
+---
+title: Debugging Incremental Builds
+---
+
+With the [release of Gatsby v3](/docs/reference/release-notes/v3.0#incremental-builds-in-oss) Incremental Builds is available to everyone. This improvement is (re)generating only subset of HTML files that need to be generated. To be able to use it you will need to keep the `.cache` and `public` directories from previous builds.
+
+## How does it work?
+
+To be able to understand why you might see more generated pages than expected it's helpful to know how Gatsby tracks "inputs" when generating HTML files. In particular Gatsby tracks:
+
+- which page template the page is using
+- results of page query
+- results of static queries used by a page template
+- frontend source code (shared and also browser `gatsby-browser.js` / SSR `gatsby-ssr.js` specifically)
+
+When those inputs change since the last build, the HTML files are marked to be regenerated. If they don't change, Gatsby can reuse HTML files generated in previous build.
```

Comment:
How much details we want to put in here? I.e. changing js files will result in all .html to be regenerated, result of page query will result in single file invalidation and result of static query will invalidate `.html` files that rely on it (it can be anywhere from single page to all pages dependinging wether static query is used/imported by all templates, single template etc)

Reply:
Ah, ok I see you put some `Expected behaviours` at the bottom - just thinking out loud - wouldn't it make sense to move that section as-is to after this sentence and before debugging to set up baseline expectations before explaining how to debug?